### DISCUSSION TITLE
GATv2 Run out of Memory on Heterogeneous Graph

### DISCUSSION THREAD

**Discussion by abdalgader-a-zz** [pyg-team/pytorch_geometric#4945] [Q&A]
Hi there 👋, 

I ran a [heterogeneous model ](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html) with 2-layers (256 -> 256, 256 - > 128) of [GATv2](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html?highlight=GATv2#torch_geometric.nn.conv.GATv2Conv). The number of model's parameters are 1.7M. It rise an OOM error.

Subgraph  is on SparseTensor format with 46,635 nodes with 128 dimension and 12M edges. 

GPU Spec:  Tesla P100 SXM2 16GB

PS: The program run normally with [SAGEConv](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html?highlight=GATv2#torch_geometric.nn.conv.SAGEConv).

Any help or suggestions ? 

Thanks!

**Comment by wsad1** [comment]
It's kind of surprising that you are getting an OOM. Could you maybe post some snippet of your code.

**Reply by rusty1s** [reply]
`GATv2Conv` will materialize the edge representation, e.g., it will compute 12M messages. One option to mitigate OOMs would be to reduce the number of edges in your graph, e.g., by edge dropout.

**Reply by abdalgader-a-zz** [reply]
Thanks @wsad1 for your replying. It ran OOM during the forward step, code is quite similar to [this](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html#:~:text=from%20torch_geometric.nn%20import%20GATConv,aggr%3D%27sum%27) with using `NeighborLoader` for sampling subgraph. Maybe the graph is the key point, as I'm using undirected `OGB-MAG`  with extra 3 edge types. 

 Thanks @rusty1s for your illustration. Would it possible to illustrate why it's not the case with `SAGEConv`? Is this due to the attention  implementation on `GATv2Conv`?

**Reply by rusty1s** [reply]
Yes, it is due to attention mechanism. We learn multiple edge weights here, which does not let us use sparse-matrix multiplication. There is an open PR that tries to eliminate this, see https://github.com/pyg-team/pytorch_geometric/pull/2380.
